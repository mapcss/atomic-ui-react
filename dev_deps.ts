export {
  anyBoolean,
  anyFunction,
  anyNumber,
  anyString,
  defineExpect,
  defineGlobalThis,
  fn,
  jestMatcherMap,
  jestModifierMap,
} from "https://deno.land/x/unitest@v1.0.0-beta.82/mod.ts";
import {
  defineExpect,
  equal,
  jestMatcherMap,
  jestModifierMap,
  MatchResult,
} from "https://deno.land/x/unitest@v1.0.0-beta.82/mod.ts";
export { FakeTime } from "https://deno.land/std@0.136.0/testing/time.ts";
export { describe, it } from "https://deno.land/std@0.136.0/testing/bdd.ts";
export { assertSnapshot } from "https://deno.land/std@0.136.0/testing/snapshot.ts";
export { default as clsx } from "https://esm.sh/clsx@1.1.1?pin=v78";
export async function setupJSDOM(): Promise<void> {
  const { JSDOM } = await import("https://esm.sh/jsdom@19.0.0?pin=v76").then((
    module,
  ) => module.default);

  const doc = new JSDOM(`<!DOCTYPE html>`);
  globalThis.document = doc.window.document;
  globalThis.HTMLIFrameElement = doc.window.HTMLIFrameElement;
  globalThis.Node = doc.window.Node;
}
export const expect = defineExpect({
  matcherMap: {
    ...jestMatcherMap,
    toHaveFocus,
    toHaveAttribute,
  },
  modifierMap: jestModifierMap,
});

function toHaveFocus(value: Element): MatchResult {
  return {
    pass: value.ownerDocument.activeElement === value,
    expected: "Actual element is focused",
  };
}

function toHaveAttribute(
  el: Element,
  name: string,
  expected: unknown,
): MatchResult {
  if (!el.hasAttribute(name)) {
    return {
      pass: false,
      expected: `[${name}] is not exists`,
    };
  }
  return {
    pass: equal(el.getAttribute(name), expected),
    expected: `${name} is ${Deno.inspect(expected)}`,
  };
}

export function setupRaf(): () => void {
  Object.defineProperty(globalThis.window, "requestAnimationFrame", {
    value: (handler: TimerHandler) => globalThis.setTimeout(handler, 1),
    configurable: true,
    writable: true,
  });

  Object.defineProperty(globalThis.window, "cancelAnimationFrame", {
    value: globalThis.clearTimeout,
    configurable: true,
    writable: true,
  });

  const reset = (): void => {
    globalThis.requestAnimationFrame = _.requestAnimationFrame;
    globalThis.cancelAnimationFrame = _.cancelAnimationFrame;
  };

  return reset;
}

// deno-lint-ignore no-explicit-any
export type ParamReturn<T extends (...args: any[]) => unknown> = [
  ...Parameters<T>,
  ReturnType<T>,
];

const _: {
  getComputedStyle: typeof window.getComputedStyle;
  requestAnimationFrame: typeof globalThis.requestAnimationFrame;
  cancelAnimationFrame: typeof cancelAnimationFrame;
} = {
  getComputedStyle: globalThis.getComputedStyle,
  requestAnimationFrame: globalThis.requestAnimationFrame,
  cancelAnimationFrame: globalThis.cancelAnimationFrame,
};

export function mockGetComputedStyle(mock: () => CSSStyleDeclaration) {
  Object.defineProperty(globalThis.window, "getComputedStyle", {
    value: mock,
    writable: true,
    configurable: true,
  });

  const reset = (): void => {
    globalThis.getComputedStyle = _.getComputedStyle;
  };

  return reset;
}
