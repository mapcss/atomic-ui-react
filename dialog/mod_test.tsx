import { Dialog, DialogDescribe, DialogProvider, DialogTitle } from "./mod.ts";
import { assertSnapshot, describe, it, setupJSDOM } from "../dev_deps.ts";
import { render } from "@testing-library/react";
import SSRProvider from "../ssr/ssr_provider.ts";

const describeTests = describe({
  name: "Dialog E2E test",
  async beforeEach() {
    await setupJSDOM();
  },
});

it(describeTests, "should be render as", (t) => {
  const { container } = render(
    <DialogProvider>
      <Dialog isShow hasDescribe hasTitle>
        <DialogTitle>Title</DialogTitle>
        <DialogDescribe>Describe it</DialogDescribe>

        <button>Cancel</button>
      </Dialog>
    </DialogProvider>,
    {
      wrapper: ({ children }) => <SSRProvider>{children as never}</SSRProvider>,
    },
  );

  assertSnapshot(t, container.innerHTML);
});
