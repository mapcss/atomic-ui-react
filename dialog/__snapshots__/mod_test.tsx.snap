export const snapshot = {};

snapshot[`Dialog E2E test > should be render as 1`] = `'<div role="dialog" aria-modal="true" aria-labelledby="atomic-ui-0-dialog-title" aria-describedby="atomic-ui-0-dialog-describe"><h3 id="atomic-ui-0-dialog-title">Title</h3><p id="atomic-ui-0-dialog-describe">Describe it</p><button>Cancel</button></div>'`;
